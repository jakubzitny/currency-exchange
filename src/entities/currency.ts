// export type Currency = 'USD' | 'GBP' | 'EUR'

export enum Currency {
  USD = 'USD',
  GBP = 'GBP',
  EUR = 'EUR',
}
