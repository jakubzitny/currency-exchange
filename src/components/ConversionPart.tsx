import classNames from 'classnames'
import React, { useState } from 'react'

import { getCurrencySymbol } from '../utils/currency-signs'
import { roundAmount } from '../utils/validation-utils'
import { Currency } from '../entities/currency'

import styles from '../styles/ConversionPart.module.css'

type Props = {
  currency: Currency
  exchangeAmount: number
  walletAmount: number
  rate: number
  rateCurrency: Currency
  type: 'source' | 'target'
  onAmountChangeRequest: (e: React.ChangeEvent<HTMLInputElement>) => void
}

const ConversionPart: React.FC<Props> = (props: Props) => {
  const { rate } = props

  return (
    <div
      className={classNames(styles.container, {
        [styles.containerTarget]: props.type === 'target',
      })}
    >
      <div className={styles.exchangeInfo}>
        <div className={styles.exchangeInfoCurrency}>{props.currency}</div>
        <div className={styles.exchangeInfoAmount}>
          <input
            className={styles.input}
            value={props.exchangeAmount}
            onChange={props.onAmountChangeRequest}
          />
        </div>
      </div>

      <div className={styles.walletInfo}>
        <div className={styles.walletInfoAmount}>
          You have {getCurrencySymbol(props.currency)}
          {props.walletAmount}
        </div>
        {props.type === 'target' && (
          <div className={styles.rate}>
            {getCurrencySymbol(props.currency)}1 ={' '}
            {getCurrencySymbol(props.rateCurrency)}
            {roundAmount(props.type === 'target' ? 1 / rate : rate)}
          </div>
        )}
      </div>
    </div>
  )
}

export default ConversionPart
