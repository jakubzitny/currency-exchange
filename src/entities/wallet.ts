import { Record } from 'immutable'
import { Currency } from './currency'

interface Defaults {
  currency: Currency
  amount: number
}

const defaults: Defaults = {
  currency: Currency.EUR,
  amount: 0,
}

export default class Wallet extends Record<Defaults>(defaults) {}
