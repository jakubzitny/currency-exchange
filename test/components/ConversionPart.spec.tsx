import chai, { expect } from 'chai'
import chaiSinon from 'sinon-chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import React from 'react'
import { spy } from 'sinon'
import { List } from 'immutable'

import { Currency } from '../../src/entities/currency'
import Wallet from '../../src/entities/wallet'
import ConversionPart from '../../src/components/ConversionPart'

chai.use(chaiJestSnapshot)
chai.use(chaiSinon)

describe('ConversionPart', () => {
  const createConversionPart = (
    props: Object = {},
    type: 'source' | 'target' = 'source'
  ) => {
    const defaultProps = {
      currency: Currency.EUR,
      exchangeAmount: 50,
      walletAmount: 123,
      rate: 1.5,
      rateCurrency: Currency.USD,
      onAmountChangeRequest: () => {},
    }
    const switcher = <ConversionPart type={type} {...defaultProps} {...props} />
    return shallow(switcher)
  }

  it('should render source ConversionPart', () => {
    const wrapper = createConversionPart()
    expect(wrapper).to.matchSnapshot()
  })

  it('should render target ConversionPart', () => {
    const wrapper = createConversionPart({}, 'target')
    expect(wrapper).to.matchSnapshot()
  })

  it('should handle input change', () => {
    const onAmountChangeRequest = spy()
    const wrapper = createConversionPart({ onAmountChangeRequest })

    wrapper.find('input').simulate('change')
    expect(onAmountChangeRequest).to.have.been.called
  })
})
