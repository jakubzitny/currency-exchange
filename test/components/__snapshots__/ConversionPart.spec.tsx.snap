// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionPart should render source ConversionPart 1`] = `ShallowWrapper {}`;

exports[`ConversionPart should render target ConversionPart 1`] = `ShallowWrapper {}`;
